src/goBrowsePackage.ts

```diff
@@ -26,64 +26,73 @@ export function browsePackages() {
+		if (!stdout || stdout.indexOf(':') === -1) {
+			if (isGoListComplete()) {
+				return showPackageList();
+			}
+
+			return showTryAgainLater();
 		}
-		showPackages(selectedText);
-	}, 1000);
 
+		let matches = stdout && stdout.match(/(.*):\[(.*)\]:\[(.*)\]:\[(.*)\]/);
+		if (matches) {
+			let dir = matches[1];
+			let files = matches[2] ? matches[2].split(' ') : [];
+			let testfiles = matches[3] ? matches[3].split(' ') : [];
+			let xtestfiles = matches[4] ? matches[4].split(' ') : [];
+			files = files.concat(testfiles);
+			files = files.concat(xtestfiles);
+			vscode.window.showQuickPick(files, { placeHolder: `Below are Go files from ${pkg}` }).then(file => {
+				// if user abandoned list, file will be null and path.join will error out.
+				// therefore return.
+				if (!file) return;
+
+				vscode.workspace.openTextDocument(path.join(dir, file)).then(document => {
+					vscode.window.showTextDocument(document);
+				});
+			});
+		}
+	});
 }
 
-function showPackages(selectedText: string) {
-	const goRuntimePath = getGoRuntimePath();
-	if (!goRuntimePath) {
-		return;
-	}
+function showPackageList() {
 	goListAll().then(pkgMap => {
 		const pkgs: string[] = Array.from(pkgMap.keys());
 		if (!pkgs || pkgs.length === 0) {
 			return vscode.window.showErrorMessage('Could not find packages. Ensure `go list all` runs successfully.');
 		}
-		let selectPkgPromise: Thenable<string> = Promise.resolve(selectedText);
-		if (!selectedText || pkgs.indexOf(selectedText) === -1) {
-			selectPkgPromise = vscode.window.showQuickPick(pkgs, { placeHolder: 'Select a package to browse' });
-		}
-		selectPkgPromise.then(pkg => {
-			cp.execFile(goRuntimePath, ['list', '-f', '{{.Dir}}:{{.GoFiles}}:{{.TestGoFiles}}:{{.XTestGoFiles}}', pkg], (err, stdout, stderr) => {
-				if (!stdout || stdout.indexOf(':') === -1) {
-					return;
-				}
-				let matches = stdout.match(/(.*):\[(.*)\]:\[(.*)\]:\[(.*)\]/);
-				if (matches) {
-					let dir = matches[1];
-					let files = matches[2] ? matches[2].split(' ') : [];
-					let testfiles = matches[3] ? matches[3].split(' ') : [];
-					let xtestfiles = matches[4] ? matches[4].split(' ') : [];
-					files = files.concat(testfiles);
-					files = files.concat(xtestfiles);
-					vscode.window.showQuickPick(files, { placeHolder: `Below are Go files from ${pkg}` }).then(file => {
-						// if user abandoned list, file will be null and path.join will error out.
-						// therefore return.
-						if (!file) return;
 
-						vscode.workspace.openTextDocument(path.join(dir, file)).then(document => {
-							vscode.window.showTextDocument(document);
-						});
-					});
-				}
+		vscode
+			.window
+			.showQuickPick(pkgs, { placeHolder: 'Select a package to browse' })
+			.then(pkgFromDropdown => {
+				if (!pkgFromDropdown) return;
+				showPackageFiles(pkgFromDropdown);
 			});
-		});
 	});
 }
 
+function showTryAgainLater() {
+	// `go list all` has not completed. Wait for a second which is an acceptable duration of delay.
```

Comment:
if at the end of this second, `go list` completes, we won't be showing anything. Was that intended?